parlai/tasks/airdialogue/agents.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from parlai.core.teachers import FixedDialogTeacher
+from .build import build
+import os
+import json
+
+
+def _path(opt):
+    # build the data if it does not exist
+    build(opt)
+
+    # set up path to data (specific to each dataset)
+    jsons_path = os.path.join(
+        opt['datapath'], 'airdialogue', 'airdialogue_data', 'airdialogue'
+    )
+    return jsons_path
+
+
+class AirDialogueTeacher(FixedDialogTeacher):
+    """
+    AirDialogue Teacher.
+
+    This also contains other files (dev_kb.json, train_kb.json) with flight data about
+    return flights, price, connections, flight airlines, departure times, and other
+    flight information.
+    """
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        jsons_path = _path(opt)
+        self.datatype = opt['datatype'].split(':')[0]
+        self.messages = []
+        self._setup_data(jsons_path)
+        self.id = 'airdialogue'
+        self.reset()
+
+    def _setup_data(self, jsons_path):
+        train_path = os.path.join(jsons_path, 'train_data.json')
+        test_valid_path = os.path.join(jsons_path, 'dev_data.json')
+        if self.datatype.startswith('test'):
+            with open(test_valid_path) as f:
+                for line in f:
+                    if len(line) > 1:
+                        self.messages.append(json.loads(line)['dialogue'])
+            self.messages = self.messages[len(self.messages) // 2 :]
+        elif self.datatype.startswith('valid'):
+            with open(test_valid_path) as f:
+                for line in f:
+                    if len(line) > 1:
+                        self.messages.append(json.loads(line)['dialogue'])
+            self.messages = self.messages[: len(self.messages) // 2]
+        else:
+            with open(train_path) as f:
+                for line in f:
+                    if len(line) > 1:
+                        self.messages.append(json.loads(line)['dialogue'])
+
+    def num_examples(self):
+        examples = 0
+        for data in self.messages:
+            examples += len(data) // 2
+        return examples
+
+    def num_episodes(self):
+        return len(self.messages)
+
+    def get(self, episode_idx, entry_idx=0):
+        log_idx = entry_idx * 2
+        entry = self.messages[episode_idx][log_idx]
+        entry = entry.split(': ')[1]
+        last_backnforth_idx = len(self.messages[episode_idx]) - 2
+        episode_done = log_idx >= last_backnforth_idx
+        if log_idx < last_backnforth_idx:
+            label_text = self.messages[episode_idx][log_idx + 1]
+            label_text = label_text.split(': ')[1]
+            labels = [label_text]
+        else:
+            labels = ['']
```

Comment:
when does this happen? i think we should avoid surfacing examples for which the label field is empty

Reply:
This happens when the first speaker is also the very last speaker, so occasionally at the very end of episodes when they have an odd length.

Reply:
in this case let's drop the last example